### PR TITLE
Validation Duration Options and Duration Formatter Stub

### PR DESCRIPTION
**Pull request by kartva**
- implement option validation algorithm ([TC39 GetDurationUnitOptions](https://tc39.es/proposal-intl-duration-format/#sec-getdurationunitoptions))
- create `DurationFormatter` stub

@younies
@sffc

### CODE REVIEW

File: components/experimental/src/duration/options.rs [hunk 1]
```diff
@@ -0,0 +1,327 @@
...
+    Numeric,
+}
+
+/// Configures the style of the second field.
+#[non_exhaustive]
+#[derive(Debug, Default, Clone, Copy, PartialEq, Eq)]
+pub enum SecondStyle {
+    #[default]
+    /// The default style for this unit. Determined by base style.
+    Default,
```
Comment (younies) [inline components/experimental/src/duration/options.rs:264]: This is can be moved to the code logic or we can call it `BaseStyle`

File: components/experimental/src/duration/formatter.rs [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+impl DurationFormatter {
+    pub fn new(options: DurationFormatterOptions) -> Result<Self, DurationFormatterError> {
+        Ok(DurationFormatter {
+            options: ValidatedDurationFormatterOptions::validate(options)?,
+        })
+    }
+}
+
+#[derive(Debug, Default, Clone, Copy, PartialEq, Eq)]
+pub struct ValidatedDurationFormatterOptions {
```
Comment (younies) [inline components/experimental/src/duration/formatter.rs:18]: add comment

File: components/experimental/src/duration/formatter.rs [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+    /// Style for nanosecond
+    nanosecond: FieldStyle,
+    /// Visibility control for nanosecond
+    nanosecond_visibility: FieldDisplay,
+
+    /// Number of fractional digits to use when formatting sub-second units (milliseconds, microseconds, nanoseconds).
+    ///  # NOTE 
+    ///       - Only takes effect when the subsecond units are styled as `Numeric`.
+    ///       - Zero means no fractional digits.
+    fractional_digits: FractionalDigits,
```
Comment (younies) [inline components/experimental/src/duration/formatter.rs:68]: I changed this slightly

File: components/experimental/src/duration/formatter.rs [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+            }
+
+            prev_style = *style;
+        }
+
+        Ok(unified)
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
```
Comment (younies) [inline components/experimental/src/duration/formatter.rs:213]: add comment

File: components/experimental/src/duration/options.rs (hunk 1 above)
Comment (DesmondWillowbrook) [reply components/experimental/src/duration/options.rs]: Changed `Style::Default` to `Option<Style>`.

File: components/experimental/src/duration/formatter/validated_options.rs [hunk 5]
```diff
@@ -0,0 +1,358 @@
...
+    nanosecond_visibility: FieldDisplay,
+
+    /// Number of fractional digits to use when formatting sub-second units (milliseconds, microseconds, nanoseconds).
+    /// ### Note:
+    /// - Only takes effect when the subsecond units are styled as `Numeric`.
+    /// - Zero means no fractional digits.
+    fractional_digits: FractionalDigits,
+}
+
+#[non_exhaustive]
```
Comment (younies) [inline components/experimental/src/duration/formatter/validated_options.rs:62]: add comment

File: components/experimental/src/duration/formatter/validated_options.rs [hunk 6]
```diff
@@ -0,0 +1,358 @@
...
+            }
+
+            prev_style = *style;
+        }
+
+        Ok(builder.try_into().unwrap())
+    }
+
+    #[allow(dead_code)]
+    pub(crate) fn iter_units(&mut self) -> [(Unit, &mut FieldStyle, &mut FieldDisplay); 10] {
```
Comment (younies) [inline components/experimental/src/duration/formatter/validated_options.rs:183]: add comment

File: components/experimental/src/duration/formatter/validated_options.rs [hunk 7]
```diff
@@ -0,0 +1,358 @@
...
+    Hour,
+    Minute,
+    Second,
+    Millisecond,
+    Microsecond,
+    Nanosecond,
+}
+
+impl Unit {
+    pub(crate) fn digital_default(&self) -> FieldStyle {
```
Comment (younies) [inline components/experimental/src/duration/formatter/validated_options.rs:344]: add comment

File: components/experimental/src/duration/options.rs [hunk 8]
```diff
@@ -73,15 +74,82 @@ pub enum FractionalDigits {
...
+pub enum FieldDisplay {
     /// Only display this field if it is non-zero.
     Auto,
     /// Always display this field.
     Always,
 }
 
+#[non_exhaustive]
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub(crate) enum FieldStyle {
```
Comment (younies) [inline components/experimental/src/duration/options.rs:87]: add comment

**Review by younies — APPROVED** [review]

**Comment by kartva** [comment]
Is there a blocker to merging this branch?

**Pull request merged**